### PR TITLE
Unify build script draft

### PR DESCRIPTION
**Pull request by SinTh0r4s**
Do we want to implement the version passing to GT_mod.java? We would need to adapt some minor things internally, then. Since GT uses hardcoded integers for versioning.

Also, when this is merged, it is likely to break some other build scripts. Which will be more work. But when that is done we could also provide gregtech with a master branch instead of experimental.

### CODE REVIEW

File: repositories.gradle [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+        name = "chickenbones"
+        url = "https://nexus.covers1624.net/repository/maven-hosted/"
+    }
+    maven {
+        name = "ic2, forestry"
+        url = "http://downloads.gtnewhorizons.com/Mods_for_Jenkins/"
+    }
+    maven { // EnderIO & EnderCore
+        name = 'tterrag Repo'
+        url = "http://maven.tterrag.com"
```
Comment (glowredman) [inline repositories.gradle:18]: Can use HTTPS

File: repositories.gradle [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    }
+    maven { // TConstruct
+        name = 'DVS1 Maven FS'
+        url = 'http://dvs1.progwml6.com/files/maven'
```
Comment (glowredman) [inline repositories.gradle:22]: Can use HTTPS

File: repositories.gradle [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+    }
+    maven { // AppleCore
+        url = "http://www.ryanliptak.com/maven/"
```
Comment (glowredman) [inline repositories.gradle:25]: Can use HTTPS

File: repositories.gradle [hunk 4]
```diff
@@ -0,0 +1,43 @@
+// Add any additional repositiroes for your dependencies here
```
Comment (glowredman) [inline repositories.gradle:1]: Typo: repositories

File: repositories.gradle [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+
+repositories {
+    maven {
+        name = "gt"
+        url = "https://gregtech.overminddl1.com/"
+    }
+    maven {
+        name = "chickenbones"
+        url = "https://nexus.covers1624.net/repository/maven-hosted/"
```
Comment (glowredman) [inline repositories.gradle:10]: Why do you use covers1624's maven instead of chickenbones' maven?

File: build.gradle [hunk 6]
```diff
@@ -1,180 +1,376 @@
...
         }
         maven {
-            name = "jitpack.io"
-            url = "https://jitpack.io"
+            name = "sonatype"
+            url = "https://oss.sonatype.org/content/repositories/snapshots/"
         }
     }
     dependencies {
         classpath 'com.github.GTNH2:ForgeGradle:FG_1.2-SNAPSHOT'
```
Comment (bombcar) [inline build.gradle:28]: Isn't this the now-out-of date one and we should be on com.github.GTNewHorizons:ForgeGradle:FG_1.2-SNAPSHOT or similar?

Comment (mitchej123) [reply build.gradle:28]: https://github.com/GTNewHorizons/ForgeGradle/releases/tag/stable-2021-11-20

File: .github/workflows/build-and-test.yml [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+      run: ./gradlew setupCIWorkspace
+    
+    - name: Build the mod
+      run: ./gradlew build
+
+    - name: Run server for 1 minute
+      run: |
+        mkdir run
+        echo "eula=true" > run/eula.txt  
+        timeout 10 ./gradlew runServer || true
```
Comment (Glease) [inline .github/workflows/build-and-test.yml:40]: Did you actually just have this same line in your template? 10 second == 1min?

File: .github/workflows/gradle.yml [hunk 8]
```diff
@@ -1,34 +0,0 @@
...
-          distribution: 'adopt'
-          cache: gradle
-      - name: Grant execute permission for gradlew
-        run: chmod +x gradlew
-      - name: Setup the workspace
-        run: ./gradlew setupCIWorkspace
-      - name: Build with Gradle
-        run: ./gradlew build
-
-  test-scala-presence:
```
Comment (Glease) [inline .github/workflows/gradle.yml:27]: This should not be just deleted. Maybe remove the build job and name this file like `test-scala-presence.yml`

**Review by Glease — CHANGES_REQUESTED** [review]
I'm skeptic at changing the version. It's an invitation to distaster. I'd imagine every single addon or weird mod that depends on gregtech in an unexpected way to break.

**Comment by bombcar** [comment]
> I'm skeptic at changing the version. It's an invitation to distaster. I'd imagine every single addon or weird mod that depends on gregtech in an unexpected way to break.

At some point we bite that bullet; but doesn't dream up the version each release?

**Comment by Glease** [comment]
> At some point we bite that bullet; but doesn't dream up the version each release?

No. What dream change is really just produced archive version. What is supplied to fml hasn't really changed for more than half a decade.

File: .github/workflows/build-and-test.yml (hunk 7 above)
Comment (Glease) [reply .github/workflows/build-and-test.yml:40]: @TheElan

File: .github/workflows/gradle.yml (hunk 8 above)
Comment (SinTh0r4s) [reply .github/workflows/gradle.yml:27]: That files is gone completely

**Comment by SinTh0r4s** [comment]
The old internal version is in code is still there, but marked deprecated. That way we won't break any addons!

File: .github/workflows/build-and-test.yml (hunk 7 above)
Comment (SinTh0r4s) [reply .github/workflows/build-and-test.yml:40]: @TheElan

**Comment by mitchej123** [comment]
He's referring to this - 
image

**Comment by bombcar** [comment]
Let me guess, code should use `modid` but will depend on `name` instead ...

**Comment by SinTh0r4s** [comment]
> the version MC1710 --> GRADLETOKEN_VERSION

You are correct! Should stick with the constant. Gonna push the change in a second!

**Comment by Glease** [comment]
Also in mcmod.info. I'm not sure which will take precedence.

**Comment by mitchej123** [comment]
I think the mcmod.info is fine; but I do agree with the GT_MOD.java -> @MOD(version) revert

**Comment by mitchej123** [comment]
My (possibly incorrect) interpretation is that launchers (mutlimc) and FML uses the mcmod.info for version detection, and other mods use the @MOD annotation version.

**Comment by glowredman** [comment]
Afaik Forge or FML also uses the version in the @Mod annotation to check dependencies

File: .github/workflows/build-and-test.yml (hunk 7 above)
Comment (TheElan) [reply .github/workflows/build-and-test.yml:40]: That's odd, it shoild be 1min so server got moretime to fail

**Comment by glowredman** [comment]
As far as I'm concerned it should be handled like this (in general):
- modid: have a constant field somewhere (e.g. main mod class or Reference/Constants/... class). @Mod refers to that, same text in mcmod.info (no placeholder). The modid should NOT be changed *ever*, so having a token for it is just redundant and error prone.
- modname: same as above, not as important to keep the exact same String, but still not necessary to have the script replace it every single time the mod is built.
- modversion: use of a token to replace version in mcmod.info and @Mod-annotation (like currently implemented)
- mcversion: could use a token - but why? we aren't going to change it, so just don't. It's essentially voided processing power.

**Comment by mitchej123** [comment]
I tend to agree with you @glowredman   I think what is happening here though is there is a template mod repo that instead of having to change it in 15 different places, it's limited to the build script.  Easier to fork and start new mods, but it does come across as a bit weird when refactoring existing build scripts.  

That being said, I'm fine with it as a part of the build script standardization effort.

**Comment by glowredman** [comment]
As far as I understand the gradle.properties file, not setting the fields will result in the script not attempting to replace them. So it would be easy to leave that part of the script as-is  and only have the version field be poulated. This way the buildscript can be used as a template, my points above are still met and it would increase the amount of work.

**Comment by TheElan** [comment]
By me having modid in token is fine, you use it in different places, beside code and to avoid any accidental inconsistencies it's better to have it in one place, and most convenient one here is `gradle.properties`

I also like the approach of keeping all metadata in one place (mod name, version, MC version, etc) it's more consistent then values scattered around the project, and processing cost here is minimal, probably a fraction of a sec when building (constants are inlined anyway).

**Comment by SinTh0r4s** [comment]
> As far as I understand the gradle.properties file, not setting the fields will result in the script not attempting to replace them. So it would be easy to leave that part of the script as-is and only have the version field be poulated. This way the buildscript can be used as a template, my points above are still met and it would increase the amount of work.

Increase it will^^
The replacement in mcmod.info is not optional in the current implementation (absolutely an oversight on my part). All 4 values must be replaced. Changing it now will create PRs to all migrated build scripts again and half a day of tedious work. I don't think this is worth the effort.

And there are upsides to it as well. As Elan said: consistency! And with 1/3 or all mods I have migrated so far the consistency was NOT given.

**Comment by glowredman** [comment]
> By me having modid in token is fine, you use it in different places, beside code and to avoid any accidental inconsistencies it's better to have it in one place, and most convenient one here is `gradle.properties`
> 
> I also like the approach of keeping all metadata in one place (mod name, version, MC version, etc) it's more consistent then values scattered around the project, and processing cost here is minimal, probably a fraction of a sec when building (constants are inlined anyway).

I disagree. Using tokens would increase the chance of errors: if you want to reference the modid/modname from anywhere, you need to copy/retype the string each time. With a constant from the Main mod calls or some type of reference class, you can just reference that. This is the way it is handled for years now and there were never problems with it. Using a constant also allows dependants to reference it, which is not possible with a token (not to mention that with the ccurrent script you can't replace the token in more than one file anyway). Having a constant holding a token would be possible but defeats the purpose. Having something so important like the modid accidentally change while converting 200+ repos is a far greater risk than anything one gains from this.
Outside of the code the only places I can think of are the mcmod.info file and the namespace of the assets folder and this namespace is not necessarily coupled to the modid anyways.


> > As far as I understand the gradle.properties file, not setting the fields will result in the script not attempting to replace them. So it would be easy to leave that part of the script as-is and only have the version field be poulated. This way the buildscript can be used as a template, my points above are still met and it would increase the amount of work.
> 
> Increase it will^^
> The replacement in mcmod.info is not optional in the current implementation (absolutely an oversight on my part). All 4 values must be replaced. Changing it now will create PRs to all migrated build scripts again and half a day of tedious work. I don't think this is worth the effort.
> 
> And there are upsides to it as well. As Elan said: consistency! And with 1/3 or all mods I have migrated so far the consistency was NOT given.

That is really unfortunate. I really appreciate the effort you guys have put into this, but imo this is just overkill and not good practice.

**Comment by TheElan** [comment]
> I disagree. Using tokens would increase the chance of errors: if you want to reference the modid/modname from anywhere, you need to copy/retype the string each time.

I agree, you shouldn't use it as a adhock string injection. The intention here is to use token only to inject value in a class with constants and reference it's members in code. (probably @SinTh0r4s injected it directly into annotation values by mistake, there should be a central constant class like https://github.com/SinTh0r4s/ExampleMod1.7.10/blob/main/src/main/java/com/myname/mymodid/Tags.java)

> Outside of the code the only places I can think of are the mcmod.info file and the namespace of the assets folder and this namespace is not necessarily coupled to the modid anyways.

There are other small things, like jar name, actual artifact name for referencing purposes, if using mixins - configuration file which is generated, etc.

**Comment by mitchej123** [comment]
It's being replaced into a constant, in the main mod file, so any other mod can continue relying on it as usual.

Having forked and maintained a bunch of mods, they all do it slightly differently and I appreciate it being standardized.

**Comment by glowredman** [comment]
> There are other small things, like jar name, actual artifact name for dependency purposes, if using mixins - configuration file which is generated, etc.

Sure but those cannot break worlds like a changed modid when registering items/blocks/entities/... can.

**Comment by mitchej123** [comment]
I do agree any replacement should be done into a constant that can be relied upon externally!

**Comment by TheElan** [comment]
> > There are other small things, like jar name, actual artifact name for dependency purposes, if using mixins - configuration file which is generated, etc.
> 
> Sure but those cannot break worlds like a changed modid when registering items/blocks/entities/... can.

Agree, that would be a disaster, and for that we have review, so people can double check that modid is the same after refactor

**Comment by Glease** [comment]
Why merge this when there is unresolved requested changes?

File: .github/workflows/gradle.yml (hunk 8 above)
Comment (Glease) [reply .github/workflows/gradle.yml:27]: What do you mean? test-scala-presence is not needed whatsoever?

**Pull request merged**